game/src/edit/traffic_signals/preview.rs

```diff
@@ -0,0 +1,119 @@
+use crate::app::App;
+use crate::game::{ChooseSomething, State, Transition};
+use crate::sandbox::{spawn_agents_around, SpeedControls, TimePanel};
+use abstutil::Timer;
+use ezgui::{
+    hotkey, Btn, Choice, Composite, EventCtx, GfxCtx, HorizontalAlignment, Key, Outcome, TextExt,
+    UpdateType, VerticalAlignment, Widget,
+};
+use geom::Duration;
+use map_model::IntersectionID;
+use std::collections::BTreeSet;
+
+// TODO Show diagram, auto-sync the phase.
+// TODO Auto quit after things are gone?
+struct PreviewTrafficSignal {
+    composite: Composite,
+    speed: SpeedControls,
+    time_panel: TimePanel,
+}
+
+impl PreviewTrafficSignal {
+    fn new(ctx: &mut EventCtx, app: &App) -> PreviewTrafficSignal {
+        PreviewTrafficSignal {
+            composite: Composite::new(Widget::col(vec![
+                "Previewing traffic signal".draw_text(ctx),
+                Btn::text_fg("back to editing").build_def(ctx, hotkey(Key::Escape)),
+            ]))
+            .aligned(HorizontalAlignment::Center, VerticalAlignment::Top)
+            .build(ctx),
+            speed: SpeedControls::new(ctx, app),
+            time_panel: TimePanel::new(ctx, app),
+        }
+    }
+}
+
+impl State for PreviewTrafficSignal {
+    fn event(&mut self, ctx: &mut EventCtx, app: &mut App) -> Transition {
+        ctx.canvas_movement();
+
+        match self.composite.event(ctx) {
+            Outcome::Clicked(x) => match x.as_ref() {
+                "back to editing" => {
+                    app.primary.clear_sim();
+                    return Transition::Pop;
+                }
+                _ => unreachable!(),
+            },
+            _ => {}
+        }
+
+        self.time_panel.event(ctx, app);
+        // TODO Ideally here reset to midnight would jump back to when the preview started?
+        if let Some(t) = self.speed.event(ctx, app, None) {
+            return t;
+        }
+        if self.speed.is_paused() {
+            Transition::Keep
+        } else {
+            ctx.request_update(UpdateType::Game);
+            Transition::Keep
+        }
+    }
+
+    fn draw(&self, g: &mut GfxCtx, _: &App) {
+        self.composite.draw(g);
+        self.speed.draw(g);
+        self.time_panel.draw(g);
+    }
+}
+
+// TODO I guess it's valid to preview without all turns possible. Some agents are just sad.
+pub fn make_previewer(
+    ctx: &mut EventCtx,
+    app: &App,
+    members: BTreeSet<IntersectionID>,
+    phase: usize,
+) -> Box<dyn State> {
```

Comment:
FYI: I got a crash when previewing random agents around the two intersections at the north east end of montlake.

```
thread 'main' panicked at 'Queue has spillover on Traversable::Lane(2093) at 00:00:01.0 -- can't draw Car #68, bound is -0.6725m. Laggy head is None. T
his is usually a geometry bug; check for duplicate roads going between the same intersections.', sim/src/mechanics/queue.rs:125:17                     
stack backtrace:                                                                                                                                       
   0: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt                                                                
   1: core::fmt::write                                                                                                                                 
   2: std::io::Write::write_fmt                                                                                                                        
   3: std::panicking::default_hook::{{closure}}                                                                                                        
   4: std::panicking::default_hook                                                                                                                     
   5: std::panicking::rust_panic_with_hook                                                                                                             
   6: rust_begin_unwind                                                                                                                                
   7: std::panicking::begin_panic_fmt                                                                                                                  
   8: sim::mechanics::queue::Queue::inner_get_car_positions                                                                                            
   9: sim::mechanics::queue::Queue::get_car_positions                                                                                                  
  10: sim::mechanics::driving::DrivingSimState::get_draw_cars_on                                                                                       
  11: <sim::sim::Sim as sim::render::GetDrawAgents>::get_draw_cars                                                                                     
  12: game::render::map::AgentCache::populate_if_needed                                                                                                
  13: game::app::App::get_renderables_back_to_front                                                                                                    
  14: game::app::App::draw                    
```

Not sure if it's related to this change or, if I'm reading the log message correctly, maybe it implies a bug in the osm source data?

Reply:
Ooh, good catch! Unrelated to this change, but I'm looking into this